### PR TITLE
TST: Add future dependency tests as a weekly CI job

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

Test future numpy versions with Matplotlib to see if anything needs to be done in the future to address deprecations
or other pending changes.

* Currently will run once per week at 03:47 UTC on Saturdays.
* Only runs on Linux and for 3.9/3.10 currently as this is mostly just a smoke test for future issues.
* I think it will upload an issue on failure to notify the repo, but that is untested. Perhaps there is a better way here that someone knows about.

Link to a run on my branch (it should only run on matplotlib/matplotlib now)
https://github.com/greglucas/matplotlib/runs/4204821100?check_suite_focus=true

### 3.9 + numpy-1.22.dev failures

```bash
FAILED lib/matplotlib/tests/test_units.py::test_plot_masked_units[png] - matp...
= 1 failed, 8352 passed, 148 skipped, 13 xfailed, 4 xpassed in 537.93s (0:08:57) =
```

### 3.10 + numpy-1.22.dev failures

```bash
FAILED lib/matplotlib/tests/test_axes.py::test_errorbar[png] - matplotlib.tes...
FAILED lib/matplotlib/tests/test_axes.py::test_errorbar[svg] - matplotlib.tes...
FAILED lib/matplotlib/tests/test_axes.py::test_errorbar[pdf] - matplotlib.tes...
FAILED lib/matplotlib/tests/test_backends_interactive.py::test_webagg - Asser...
FAILED lib/matplotlib/tests/test_units.py::test_plot_masked_units[png] - matp...
FAILED lib/matplotlib/tests/test_streamplot.py::test_direction[png] - matplot...
FAILED lib/mpl_toolkits/tests/test_axisartist_grid_helper_curvelinear.py::test_axis_direction[png]
FAILED lib/mpl_toolkits/tests/test_mplot3d.py::test_trisurf3d[png] - matplotl...
FAILED lib/mpl_toolkits/tests/test_mplot3d.py::test_stem3d[png] - matplotlib....
= 9 failed, 8344 passed, 148 skipped, 13 xfailed, 4 xpassed, 4 warnings in 414.94s (0:06:54) =
```

### Additional things to consider
Perhaps we should also consider adding a nightly cibuildwheel to that repository for others to test out Matplotlib and report back to us easier?

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: .github/workflows/tests-future-deps.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+          # All dependencies must have been pre-installed, so that the minver
+          # constraints are held.
+          python -m pip install --no-deps -e .
+
+      - name: Run pytest
+        run: |
+          xvfb-run -a python -mpytest -raR -n auto \
+            --maxfail=50 --timeout=300 --durations=25 --log-level=DEBUG \
+            -W error::UserWarning
```
Comment (greglucas) [inline .github/workflows/tests-future-deps.yml:144]: We actually would like to see the DeprecationWarning's as well (`-W error` without the specific UserWarning), but on Python 3.10 Numpy's testing import causes deprecation warnings due to the distutils future removal in Python 3.12. So, we would have to wait for Numpy's removal of distutils before we did that. Unless I'm missing a way to catch a specific string of deprecation warnings so we could ignore only that specific warning.

**Comment by tacaswell** [comment]
I think some of the 3.10 + numpy-1.22.dev failures are fixed on numpy's default branch (at one point I was seeing issuse with distutils warnings, but they are gone for me locally).

I am in principle 👍🏻 👍🏻  on this idea!

File: .github/workflows/tests-future-deps.yml (hunk 1 above)
Comment (QuLogic) [reply .github/workflows/tests-future-deps.yml:144]: Sure, you can be more specific with warnings filters: https://docs.python.org/3/library/warnings.html#describing-warning-filters

Comment (greglucas) [reply .github/workflows/tests-future-deps.yml:144]: Thanks for pointing that out. I tried a few things out and it was more complicated that I hoped.
See here for other discussion: https://bugs.python.org/issue43862
> While the Python API warnings.filterwarnings(action, message="", ...) uses the message as a regular expression, -W and PYTHONWARNINGS require to match *exactly* the *whole* message.

I got one match, and then another import failed, so not sure how many options/messages we want to match on if we do. I couldn't figure out how to use the "module" piece either, I thought that could refer to errors raise from `distutils` directly, or even by putting `numpy.testing` in there for where the warnings were arising from, but neither of those worked for me...

In summary, I'd leave it with the UserWarning -> Error for now and deal with DeprecationWarnings after Numpy fixes their distutils  imports.

**Comment by greglucas** [comment]
Looks like the test suite  is passing with the nightly pandas and numpy wheels now.

https://github.com/greglucas/matplotlib/runs/4290939395?check_suite_focus=true

**Comment by greglucas** [comment]
Some of the tests seem to be a little flaky when they are run: https://github.com/greglucas/matplotlib/actions/runs/1495511858
The new changes seem to be good, but `test_errorbar` is failing, even though it passed on a previous run with the same dependency versions installed...

**Comment by greglucas** [comment]
I pushed a local copy of this up to my fork: https://github.com/greglucas/matplotlib/runs/4621007693?check_suite_focus=true
and both 3.9 and 3.10 with latest numpy is failing. However, when I try to run this all locally (installing numpy from nightly wheels) the tests all pass for me. There must be something about the runner environment that makes a few of the tests flaky, but I haven't been able to figure it out yet looking through the logs.

**Comment by QuLogic** [comment]
The result images should be obtainable on the Summary page. It doesn't look like the failures are anywhere other than test images.

**Comment by greglucas** [comment]
Yep, I did download them and couldn't find any obvious reason for the failures (they are all quite small). The largest one `test_errorbar` has an expanded y-limit, causing the number of ticks to be increased, but the data in the plot appears to be correct. My initial guess is that the failures are due to floating point discrepancies, causing a ylim to be expanded to the next decade, but I'm not sure where that would be coming from since I can't reproduce locally.

**Comment by greglucas** [comment]
Putting some more notes down here:
This test suite does pass CI if I don't upgrade numpy to the nightly wheel: https://github.com/greglucas/matplotlib/actions/runs/1617303489 (I was worried I'd updated some other config to make a standard run fail). I also tried pip installing nightly numpy before everything else and after everything else (including matplotlib), to make sure it wasn't affecting the compilation of any dependencies and it failed in every case there as well.

I can't reproduce these failures on a local macosx or ubuntu machine either, so my debugging capability here is limited unfortunately.

**Comment by greglucas** [comment]
Perhaps this is related to AVX512 instructions now being available in the wheel...
https://numpy.org/devdocs/release/1.22.0-notes.html#vectorize-umath-module-using-avx-512

On the 3.8-3.10 runners:
```
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2,AVX512F,AVX512CD,AVX512_SKX
    not found = AVX512_KNL,AVX512_KNM,AVX512_CLX,AVX512_CNL,AVX512_ICL
```

On the 3.7 runner (numpy 1.21):
```
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_KNL,AVX512_KNM,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
```

My local Ubuntu is on a chip without AVX512 instructions, which could explain why I'm not able to reproduce this locally. If someone does have a system with a chip that supports AVX512 and could test locally that could verify this guess: `python -c "import numpy; numpy.show_config()"`
or if someone knows of a way to disable that instruction set from use within numpy at runtime rather than build time, I could push that up to the runner as well.

**Comment by dstansby** [comment]
> Perhaps this is related to AVX512 instructions now being available in the wheel... https://numpy.org/devdocs/release/1.22.0-notes.html#vectorize-umath-module-using-avx-512

I suspect this is correct - I'm seeing similar small floating point differences on another project that seems to be related to both numpy version and the specific machine CI is being run on.

**Comment by greglucas** [comment]
I decided to rework this and incorporate it into the current tests.yml file, which seems better than trying to maintain two copies of the CI file. There are new floating-point issues that cropped up again in Normalize with longdouble. It must be a new numpy optimization in `.min()`, as specific item-access works, so I've updated the tests to account for these changes.

**Review by jklymak — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -8,6 +8,9 @@ on:
   pull_request:
     branches-ignore:
       - v[0-9]+.[0-9]+.[0-9x]+-doc
+  schedule:
+    # 3:47 UTC on Saturdays
+    - cron: "47 3 * * 6"
```
[the remark this answers is not in the document]
Comment (greglucas) [reply .github/workflows/tests.yml:14]: Added a comment now.

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/tests/test_colors.py [hunk 3]
```diff
@@ -568,14 +568,15 @@ def test_Normalize():
...
+    norm = mcolors.Normalize(vals[0], vals[1])
+    assert norm(vals).dtype == np.longdouble
+    assert_array_equal(norm(vals), [0, 1])
     # and for scalar ones.
     eps = np.finfo(np.longdouble).resolution
     norm = plt.Normalize(1, 1 + 100 * eps)
     # This returns exactly 0.5 when longdouble is extended precision (80-bit),
     # but only a value close to it when it is quadruple precision (128-bit).
-    assert 0 < norm(1 + 50 * eps) < 1
+    np.testing.assert_array_almost_equal_nulp(norm(1 + 50 * eps), 0.5)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_colors.py:579]: This seems to be a bit too strict now, as it is failing on aarch64, ppc64le, and s390x, where the result is 0.50096339.

Also, on my 64-bit AMD system, which is apparently using `np.float128` for `np.longdouble` (though I don't know if that means 80-bit internally), it seems to return 0.5 exactly, which seems the opposite of the comment.
